Tests/ProcessLib/TestJacobianAssembler.cpp

```diff
@@ -0,0 +1,577 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include <limits>
+#include <random>
+#include <gtest/gtest.h>
+
+#include "MathLib/LinAlg/Eigen/EigenMapTools.h"
+#include "ProcessLib/LocalAssemblerInterface.h"
+#include "ProcessLib/AnalyticalJacobianAssembler.h"
+#include "ProcessLib/CentralDifferencesJacobianAssembler.h"
+
+//! Fills a vector with values whose absolute value is between \c abs_min and
+//! \c abs_max.
+void fillRandomlyConstrainedAbsoluteValues(std::vector<double>& xs,
+                                           double const abs_min,
+                                           double const abs_max)
+{
+    std::random_device rd;
+    std::mt19937 random_number_generator(rd());
+    double const abs_range = abs_max - abs_min;
+    std::uniform_real_distribution<double> rnd(abs_min,
+                                               abs_min + 2.0 * abs_range);
+
+    for (auto& x : xs) {
+        // v in [ abs_min, abs_min + 2 abs_range ]
+        auto v = rnd(random_number_generator);
+        if (v > abs_max) {
+            // produce negative values
+            // (v - abs_range) in [ abs_min, abs_max ]
+            v = -(v - abs_range);
+        }
+        x = v;
+    }
+}
+
+struct MatDiagX
+{
+    // M = diag(x1, x2, x3, ...)
+    static void getMat(std::vector<double> const& x_data,
+                       std::vector<double>& mat_data)
+    {
+        auto mat =
+            MathLib::createZeroedMatrix(mat_data, x_data.size(), x_data.size());
+        mat.diagonal().noalias() =
+            MathLib::toVector<Eigen::VectorXd>(x_data, x_data.size());
+    }
+
+    // dM/dx * y = diag(y1, y2, y3, ...)
+    static void getDMatDxTimesY(std::vector<double> const& x_data,
+                                std::vector<double> const& y_data,
+                                std::vector<double>& dMdxTy_data)
+    {
+        auto dMdxTy =
+            MathLib::createZeroedMatrix(dMdxTy_data, x_data.size(), x_data.size());
```

Comment:
When using y_data.size() instead of x_data.size(), the function becomes identical to the above, isn't it? Or are the x_data and y_data sizes different?